File: TRAQS Scheduling/TRAQS Scheduling/Views/SharedComponents.swift
import SwiftUI

// MARK: - TRAQS Loading Overlay
// Full-screen, TRAQS-styled blocking loading indicator. Shown during an async
// action that would otherwise leave the screen looking frozen (clock in/out).
// Driven by a non-nil message; the call site animates it in/out.
struct TRAQSLoadingOverlay: View {
    let message: String
    /// The action landed. Swaps the spinner for the checkmark — see
    /// `ClockProgressMark`. The caller holds this for a beat before clearing the
    /// overlay, so the confirmation is actually seen.
    var done: Bool = false
    var body: some View {
        ZStack {
            Color.black.opacity(0.45)
                .ignoresSafeArea()
            VStack(spacing: 16) {
                ClockProgressMark(done: done)
                Text(message)
                    .font(.system(size: 15, weight: .semibold))
                    .foregroundStyle(Color(hex: T.ink))
                    .contentTransition(.opacity)
            }
            .padding(.vertical, 30)
            .padding(.horizontal, 44)
            .glassSurface(in: RoundedRectangle(cornerRadius: T.cornerLg, style: .continuous), rim: true)
            .shadow(color: .black.opacity(0.28), radius: 26, y: 12)
        }
        // Swallow taps so the underlying screen can't be poked mid-action.
        .contentShape(Rectangle())
        .onTapGesture { }
    }
}

// MARK: - Clock progress mark
//
// A spinner while the request is out, then a checkmark that draws itself on.
// ONE view, shared by the PIN pad and the full-screen loading overlay, so the
// two endings of a clock in/out can't drift into different confirmations.

struct ClockProgressMark: View {
    let done: Bool
    var size: CGFloat = 64

    /// How much of the tick is drawn — animated from 0 so the stroke travels
    /// rather than appearing whole.
    @State private var trim: CGFloat = 0
    /// The disc's settle. Starts under 1 so the mark lands with a little weight.
    @State private var pop: CGFloat = 0.7
    /// The sweep's rotation, driven forever while the request is out.
    @State private var spin: Double = 0

    /// Ring weight. Proportional to `size` so the mark scales as one object.
    private var stroke: CGFloat { size * 0.085 }

    var body: some View {
        ZStack {
            if done {
                Circle()
                    .fill(Color(hex: T.statusFinished).opacity(0.16))
                    .scaleEffect(pop)
                CheckmarkPath()
                    .trim(from: 0, to: trim)
                    .stroke(Color(hex: T.statusFinished),
                            style: StrokeStyle(lineWidth: stroke,
                                               lineCap: .round, lineJoin: .round))
                    .padding(size * 0.3)
                    .scaleEffect(pop)
            } else {
                // A brand-coloured sweep on its own faint track, NOT the stock
                // `ProgressView`. The system spinner is a grey pinwheel that
                // belongs to no part of this app — next to a glass panel and a
                // gradient CTA it read as a placeholder. The track keeps the
                // circle whole so the sweep travels around something.
                Circle()
                    .stroke(Color(hex: T.accentGradientStart).opacity(0.13),
                            lineWidth: stroke)
                    .padding(stroke / 2)
                Circle()
                    .trim(from: 0, to: 0.3)
                    .stroke(
                        AngularGradient(
                            colors: [Color(hex: T.accentGradientEnd),
                                     Color(hex: T.accentGradientStart)],
                            center: .center),
                        style: StrokeStyle(lineWidth: stroke, lineCap: .round))
                    .padding(stroke / 2)
                    // Starts at 12 o'clock rather than 3, so the head leads from
                    // the top on the first frame.
                    .rotationEffect(.degrees(-90 + spin))
            }
        }
        .frame(width: size, height: size)
        .onAppear {
            withAnimation(.linear(duration: 0.85).repeatForever(autoreverses: false)) {
                spin = 360
            }
        }
        .onChange(of: done, initial: true) { _, landed in
            guard landed else { return }
            trim = 0
            pop = 0.7
            withAnimation(.spring(response: 0.34, dampingFraction: 0.6)) { pop = 1 }
            // Slightly behind the disc, so the tick draws INTO a circle that is
            // already there rather than racing it.
            withAnimation(.easeOut(duration: 0.28).delay(0.06)) { trim = 1 }
        }
        .sensoryFeedback(.success, trigger: done)
    }
}

/// The tick, as a path so it can be trimmed and drawn on.
struct CheckmarkPath: Shape {
    func path(in rect: CGRect) -> Path {
        var p = Path()
        p.move(to: CGPoint(x: rect.minX, y: rect.midY))
        p.addLine(to: CGPoint(x: rect.minX + rect.width * 0.36, y: rect.maxY))
        p.addLine(to: CGPoint(x: rect.maxX, y: rect.minY))
        return p
    }
}

// MARK: - TRAQS Wordmark (image asset)
// Uses the official brand wordmark PNG from Assets.xcassets. `size` is the
// rendered HEIGHT in points; aspect ratio is preserved.
// Light theme → "TRAQSLogoBlack" (dark ink wordmark for light surfaces).
// Dark theme  → "TRAQSLogoWhite" (paper wordmark for dark surfaces).

struct TRAQSWordmark: View {
    @Environment(ThemeSettings.self) private var themeSettings

    /// Aspect of the source art (traqs-wordmark-bold-ink-2348x1200).
    ///
    /// The frame below is deliberately DEFINITE in both axes. Constraining only
    /// the height leaves a `.resizable()` image with no minimum intrinsic width,
    /// so it is the first thing a tight HStack compresses — and because
    /// `scaledToFit` preserves the ratio, losing width also loses HEIGHT. That is
    /// how adding one wider control to the Jobs header silently shrank the logo on
    /// every page: the wordmark, not the row, absorbed the overflow. With a
    /// definite frame it holds its size and the layout has to give elsewhere.
    static let aspect: CGFloat = 2348.0 / 1200.0

    var size: CGFloat = 44
    /// Overrides the theme-driven variant for callers whose backdrop ISN'T the
    /// page colour. `true` → the black mark (light backdrop), `false` → white.
    /// The splash uses this: it sits on the accent-coloured liquid wash, not on
    /// the theme background, so `isLightTheme` is the wrong question there.
    var onLightBackground: Bool? = nil

    var body: some View {
        Image((onLightBackground ?? themeSettings.isLightTheme) ? "TRAQSLogoBlack" : "TRAQSLogoWhite")
            .resizable()
            // .interpolation(.high) tells SwiftUI to use high-quality
            // sampling when scaling the 3840-wide source down to display
            // size, and .antialiased smooths the resulting glyph edges so
            // the wordmark doesn't look pixelated in the nav header.
            .interpolation(.high)
            .antialiased(true)
            .scaledToFit()
            // Rounded to whole points: a fractional width puts the bars mark
            // beside it — and every trailing control after it — on subpixel
            // coordinates, which is visible as the header twitching a point as
            // views come and go.
            .frame(width: (size * Self.aspect).rounded(), height: size)
    }
}

// Back-compat: a few places still call TRAQSNavLogo() — keep as the image wordmark.
struct TRAQSNavLogo: View {
    var body: some View { TRAQSWordmark(size: 44) }
}

// MARK: - TRAQS Bars Mark (native, theme-tracking)
// The TRAQS "bars" lockup, drawn in SwiftUI instead of the fixed-color raster
// (`Image("TRAQSIconBars")`) so it follows the system theme: the accent bar
// tracks the user's chosen accent, and the three grey bars use the theme's
// muted ink so they stay legible on light AND dark backgrounds. `size` is the
// rendered HEIGHT in points; width is derived from the original 184×150 art.
struct TRAQSBarsMark: View {
    @Environment(ThemeSettings.self) private var themeSettings
    var size: CGFloat = 22

    // Bar widths as a fraction of the mark's full width, top → bottom, measured
    // from the original artwork. The 3rd (full-width) bar is the accent bar.
    private let widths: [CGFloat] = [0.554, 0.788, 1.0, 0.451]
    private let accentIndex = 2

    var body: some View {
        // Read themeSettings.accent / .bgPresetId so the mark re-renders live
        // when the customizer changes the accent or the light/dark background.
        let accentColor = Color(hex: themeSettings.accent)
        let _ = themeSettings.bgPresetId
        let greyColor = Color(hex: T.muted)

        let aspect: CGFloat = 184.0 / 150.0
        let fullWidth = size * aspect
        let barH = size * (27.0 / 150.0)
        let gap  = size * (14.0 / 150.0)

        VStack(alignment: .leading, spacing: gap) {
            ForEach(widths.indices, id: \.self) { i in
                RoundedRectangle(cornerRadius: barH * 0.32, style: .continuous)
                    .fill(i == accentIndex ? accentColor : greyColor)
                    .frame(width: fullWidth * widths[i], height: barH)
            }
        }
        .frame(width: fullWidth, height: size, alignment: .leading)
    }
}

// MARK: - TRAQS Header Logo (the "traqs=" lockup)
// The brand wordmark with the four-bar accent mark riding right after it like a
// trailing "=" — the same lockup the old side drawer used, now the fixed
// leading element of every page's header. Parametric so the hand-tuned bar
// alignment (offsets measured at wordmark height 64) scales cleanly to any
// header size.
struct TRAQSHeaderLogo: View {
    /// Rendered HEIGHT of the wordmark in points.
    var size: CGFloat = 44

    var body: some View {
        // The bars mark as a trailing "=" — scaled from the tuned size-64 lockup:
        // bars 21/64, pulled left 15/64 to clear the wordmark PNG's built-in right
        // padding, nudged up 1/64 to sit level with the letters.
        //
        // The pull stays a FRACTION of size rather than a flat point value so the
        // lockup holds together at every size it's rendered at (60 in the nav
        // header, 44 via TRAQSNavLogo).
        //
        // The pull is NEGATIVE SPACING, not an x-offset. An offset is visual only,
        // so the lockup used to claim ~14pt of layout width at its right edge that
        // it never drew into — phantom width taken straight out of the header's
        // budget for its trailing controls, and a big part of why one wider control
        // there was enough to start squeezing the logo. Now the lockup measures
        // exactly as wide as it looks.
        HStack(spacing: -size * (15.0 / 64.0)) {
            TRAQSWordmark(size: size)
            TRAQSBarsMark(size: size * (21.0 / 64.0))
                .offset(y: -size * (1.0 / 64.0))
        }
    }
}

// MARK: - Screen Header
// Logo lockup on the left (every page), trailing controls on the right.
// No center title — the tab bar tells the user where they are.

struct TRAQSNavHeader<Trailing: View>: View {
    /// Back-compat only: the logo now rides in the header on every page, so this
    /// flag is ignored. Kept so existing call sites keep compiling.
    var showLogo: Bool = false
    @ViewBuilder var trailing: () -> Trailing

    /// Rendered height of the logo lockup, and therefore of the header row.
    /// A computed property, not a stored one — TRAQSNavHeader is generic over its
    /// trailing content, and generic types can't hold static stored properties.
    private static var logoHeight: CGFloat { 60 }

    var body: some View {
        // Logo on the left, trailing controls on the right. Extra top padding
        // drops the whole header (and page title) down from the status bar for
        // more breathing room.
        HStack(alignment: .center, spacing: 10) {
            TRAQSHeaderLogo(size: Self.logoHeight)
                .offset(x: -13)   // nudge the lockup toward the leading edge
            Spacer()
            HStack(spacing: 6) {
                trailing()
            }
        }
        // PINNED height. Every trailing control is ≤38pt so the 60pt logo already
        // decided this row's height — but pinning it means it STAYS decided: a
        // control appearing as data loads (the approvals button, the admin
        // availability button), a badge, or anything added later can no longer
        // change the header's height and shift the page title under it by a point.
        .frame(height: Self.logoHeight)
        .padding(.horizontal, 16)
        .padding(.top, 22)
        .padding(.bottom, 12)
    }
}

extension TRAQSNavHeader where Trailing == EmptyView {
    init() { self.trailing = { EmptyView() } }
    /// Back-compat: older call sites used a centered tab name. Per the wireframes
    /// there's no centered title now — the tab bar tells the user where they are.
    /// We accept and ignore the argument so the build keeps moving while screens
    /// are rewritten.
    init(tabName _: String) { self.trailing = { EmptyView() } }
}

// MARK: - FastTRAQSPillButton — kept for back-compat
struct FastTRAQSPillButton: View {
    var body: some View {
        TIconView(icon: .sparkle, size: 15, color: Color(hex: T.sky))
            .padding(8)
            .background(Circle().fill(Color(hex: T.sky).opacity(0.12)))
            .overlay(Circle().stroke(Color(hex: T.sky).opacity(0.3), lineWidth: 1))
    }
}

// MARK: - Legacy view-modifiers (kept compiling for older call sites — the new design
// uses SBox / PillBtn / IconBtn directly).

extension View {
    func traqsCard(radius: CGFloat = T.cornerMd) -> some View {
        self
            .background(RoundedRectangle(cornerRadius: radius, style: .continuous).fill(Color(hex: T.surface)))
            .overlay(RoundedRectangle(cornerRadius: radius, style: .continuous).stroke(Color(hex: T.hair), lineWidth: 1))
            .shadow(color: Color.black.opacity(T.raisedShadowOpacity),
                    radius: T.raisedShadowRadius, x: 0, y: T.raisedShadowY)
    }

    func traqsToolbar(radius: CGFloat = T.cornerLg) -> some View {
        self
            .background(RoundedRectangle(cornerRadius: radius, style: .continuous).fill(Color(hex: T.surface)))
            .shadow(color: Color.black.opacity(T.raisedShadowOpacity),
                    radius: T.raisedShadowRadius, x: 0, y: T.raisedShadowY)
    }

    func traqsField(radius: CGFloat = T.cornerMd) -> some View {
        self
            .background(RoundedRectangle(cornerRadius: radius, style: .continuous).fill(Color(hex: T.surface)))
            .overlay(RoundedRectangle(cornerRadius: radius, style: .continuous).stroke(Color(hex: T.hair), lineWidth: 1))
    }

    /// Soft top edge for a page's ScrollView — content fades into the header
    /// instead of cutting off on a hard line. Used app-wide for a consistent
    /// fading header. Pages give their title ~`pageTitleTopInset` of top padding
    /// so it stays crisp at rest (the fade band is the top ~4.5% of the scroll).
    func topFadeMask() -> some View {
        // No-op. Every fade we tried fought the gradient background: a
        // full-height `.mask` left a hard bottom edge ("footer"), and a flat
        // top overlay left a hard top line (a flat color can't match the
        // gradient). The page title now scrolls inside the content, so there's
        // no sticky-header seam to hide at rest — so we don't fade at all.
        // Kept as a no-op so the call sites don't need to churn.
        self
    }
}

/// Small top breathing room above each page's title, just below the header
/// (no more fade band to clear, so this is tight on purpose).
let pageTitleTopInset: CGFloat = 6

File: traqs-android/app/src/main/java/com/matrixsystems/traqs/ui/screens/SharedComponents.kt
package com.matrixsystems.traqs.ui.screens

import androidx.compose.foundation.*
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.AutoAwesome
import androidx.compose.material.icons.filled.Menu
import androidx.compose.material3.*
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.foundation.BorderStroke
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.foundation.layout.offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.matrixsystems.traqs.R
import com.matrixsystems.traqs.models.JobStatus
import com.matrixsystems.traqs.models.Priority
import com.matrixsystems.traqs.ui.theme.TRAQSColors
import com.matrixsystems.traqs.ui.theme.parseColor
import com.matrixsystems.traqs.ui.theme.glassControl
import com.matrixsystems.traqs.ui.theme.TRadius
import com.matrixsystems.traqs.ui.theme.TIcons
import com.matrixsystems.traqs.ui.theme.traQSColors
import java.text.SimpleDateFormat
import java.util.*

@Composable
fun StatusBadge(status: JobStatus) {
    val c = traQSColors
    val color = status.toColor(c)
    Text(
        text = status.label,
        fontSize = 10.sp,
        fontWeight = FontWeight.Bold,
        color = color,
        modifier = Modifier
            .background(color.copy(alpha = 0.13f), RoundedCornerShape(TRadius.xs))
            .padding(horizontal = 7.dp, vertical = 3.dp)
    )
}

@Composable
fun PriorityDot(priority: Priority) {
    val c = traQSColors
    val color = priority.toColor(c)
    Box(
        modifier = Modifier
            .size(8.dp)
            .clip(CircleShape)
            .background(color)
    )
}

@Composable
fun FilterChip(
    label: String,
    isSelected: Boolean,
    color: Color = traQSColors.accent,
    onClick: () -> Unit
) {
    val c = traQSColors
    Button(
        onClick = onClick,
        shape = RoundedCornerShape(TRadius.md),
        colors = ButtonDefaults.buttonColors(
            containerColor = if (isSelected) color.copy(alpha = 0.15f) else c.surface,
            contentColor = if (isSelected) color else c.text.copy(alpha = 0.6f)
        ),
        contentPadding = PaddingValues(horizontal = 12.dp, vertical = 6.dp),
        modifier = Modifier.height(32.dp)
    ) {
        Text(text = label, fontSize = 11.sp, fontWeight = FontWeight.Bold)
    }
}

@Composable
fun SaveStatusBanner(saveStatus: com.matrixsystems.traqs.services.SaveStatus) {
    val c = traQSColors
    when (saveStatus) {
        is com.matrixsystems.traqs.services.SaveStatus.Saving -> {
            Row(
                horizontalArrangement = Arrangement.spacedBy(6.dp),
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier
                    .background(c.surface.copy(alpha = 0.9f), RoundedCornerShape(TRadius.md))
                    .padding(horizontal = 12.dp, vertical = 6.dp)
            ) {
                CircularProgressIndicator(modifier = Modifier.size(14.dp), color = Color.White, strokeWidth = 2.dp)
                Text("Saving…", fontSize = 12.sp, color = Color.White)
            }
        }
        is com.matrixsystems.traqs.services.SaveStatus.Saved -> {
            Row(
                horizontalArrangement = Arrangement.spacedBy(6.dp),
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier
                    .background(c.surface.copy(alpha = 0.9f), RoundedCornerShape(TRadius.md))
                    .padding(horizontal = 12.dp, vertical = 6.dp)
            ) {
                Icon(TIcons.Check, null, tint = c.statusFinished, modifier = Modifier.size(12.dp))
                Text("Saved", fontSize = 12.sp, color = c.text)
            }
        }
        else -> {}
    }
}

fun String.shortDate(): String {
    return try {
        val parser = SimpleDateFormat("yyyy-MM-dd", Locale.US)
        val fmt = SimpleDateFormat("M/d/yy", Locale.US)
        fmt.format(parser.parse(this)!!)
    } catch (_: Exception) { this }
}

fun com.matrixsystems.traqs.services.SaveStatus.isSaving() = this is com.matrixsystems.traqs.services.SaveStatus.Saving

fun JobStatus.toColor(c: TRAQSColors): Color = when (this) {
    JobStatus.NOT_STARTED -> c.statusNotStarted
    JobStatus.PENDING -> c.statusPending
    JobStatus.IN_PROGRESS -> c.statusInProgress
    JobStatus.ON_HOLD -> c.statusOnHold
    JobStatus.FINISHED -> c.statusFinished
}

fun Priority.toColor(c: TRAQSColors): Color = when (this) {
    Priority.LOW -> c.priLow
    Priority.MEDIUM -> c.priMedium
    Priority.HIGH -> c.priHigh
}

// The brand wordmark. Same art the iOS app ships
// (traqs-wordmark-bold-ink / -paper, 2348x1200) so the two platforms are
// literally the same mark, not two renderings of it.
//
// `size` is the rendered HEIGHT; the frame is DEFINITE in both axes on purpose.
// Constrained on height alone the image has no minimum intrinsic width, so it is
// the first thing a tight header row compresses — and since the ratio is
// preserved, losing width also loses height. That is how one wider header
// control silently shrinks the logo on every page.
private const val WORDMARK_ASPECT = 2348f / 1200f

@Composable
fun TRAQSLogo(height: Dp = 28.dp, modifier: Modifier = Modifier, useDefaultSize: Boolean = true) {
    val c = traQSColors
    val logoRes = if (c.isLight) R.drawable.traqs_logo else R.drawable.traqs_logo_white
    Image(
        painter = painterResource(logoRes),
        contentDescription = "TRAQS",
        contentScale = ContentScale.Fit,
        modifier = if (useDefaultSize) modifier.height(height).width(height * WORDMARK_ASPECT) else modifier
    )
}

// Which view the Jobs tab is showing. iOS calls this AppNav.jobsMode; the Jobs
// tab subsumed the old Schedule tab there, and the header toggle is how you get
// between them.
enum class JobsMode { LIST, GANTT }

// The Jobs header's view toggle. Names the CURRENT view rather than the one the
// tap leads to — the two glyphs alone did not say which way the tap would go.
//
// FIXED width, sized for the wider label: the Jobs header is deliberately stable
// across a mode flip, so the pill must not change width when the label does.
@Composable
fun JobsViewToggle(mode: JobsMode, onToggle: (JobsMode) -> Unit) {
    val c = traQSColors
    val isList = mode == JobsMode.LIST
    Box(
        modifier = Modifier
            .width(76.dp)
            .height(38.dp)
            .glassControl()
            .clickable { onToggle(if (isList) JobsMode.GANTT else JobsMode.LIST) },
        contentAlignment = Alignment.Center
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(4.dp)
        ) {
            Icon(
                if (isList) TIcons.Jobs else TIcons.Calendar,
                contentDescription = null,
                tint = c.text,
                modifier = Modifier.size(13.dp)
            )
            Text(
                if (isList) "List" else "Plan",
                fontSize = 11.sp, fontWeight = FontWeight.Bold, color = c.text,
            )
        }
    }
}

// A person's avatar: their profile photo if they have one, else their initials
// on their own colour. Mirrors iOS Avatar.
//
// `image` is a data: URL / base64 blob carried on the person record — the same
// shape iOS decodes — so there is no network fetch here and no image library to
// pull in. A malformed blob falls through to initials rather than showing a
// broken box.
@Composable
fun Avatar(person: com.matrixsystems.traqs.models.Person?, size: Dp = 38.dp) {
    val c = traQSColors
    val bg = remember(person?.color) {
        runCatching { parseColor(person?.color ?: "#7c3aed") }.getOrDefault(Color(0xFF7C3AED))
    }
    val bitmap = remember(person?.image) { decodeAvatar(person?.image) }

    Box(
        modifier = Modifier.size(size).clip(CircleShape).background(bg),
        contentAlignment = Alignment.Center
    ) {
        if (bitmap != null) {
            Image(
                bitmap = bitmap,
                contentDescription = person?.name,
                contentScale = ContentScale.Crop,
                modifier = Modifier.size(size).clip(CircleShape)
            )
        } else {
            val initials = (person?.name ?: "")
                .split(" ").filter { it.isNotBlank() }.take(2)
                .mapNotNull { it.firstOrNull()?.uppercaseChar() }
                .joinToString("")
            Text(
                initials.ifEmpty { "?" },
                fontSize = (size.value * 0.36f).sp,
                fontWeight = FontWeight.Bold,
                color = Color.White,
            )
        }
    }
}

private fun decodeAvatar(data: String?): androidx.compose.ui.graphics.ImageBitmap? {
    if (data.isNullOrBlank()) return null
    return runCatching {
        // Accept both a bare base64 payload and a full data: URL.
        val payload = data.substringAfter("base64,", data)
        val bytes = android.util.Base64.decode(payload, android.util.Base64.DEFAULT)
        android.graphics.BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
            ?.asImageBitmap()
    }.getOrNull()
}

// The TRAQS "bars" mark — four stacked bars, the third one the accent. Drawn in
// Compose rather than shipped as a raster so it follows the customizer's accent
// and the light/dark preset, exactly as iOS TRAQSBarsMark does.
@Composable
fun TRAQSBarsMark(size: Dp) {
    val c = traQSColors
    // Bar widths as a fraction of the mark's full width, top to bottom, measured
    // from the original artwork. The third (full-width) bar is the accent bar.
    val widths = listOf(0.554f, 0.788f, 1.0f, 0.451f)
    val accentIndex = 2
    val aspect = 184f / 150f
    val fullWidth = size * aspect
    val barH = size * (27f / 150f)
    val gap = size * (14f / 150f)

    Column(
        modifier = Modifier.width(fullWidth).height(size),
        verticalArrangement = Arrangement.spacedBy(gap)
    ) {
        widths.forEachIndexed { i, w ->
            Box(
                Modifier
                    .width(fullWidth * w)
                    .height(barH)
                    .clip(RoundedCornerShape(barH * 0.32f))
                    .background(if (i == accentIndex) c.accent else c.muted)
            )
        }
    }
}

// The "traqs=" lockup: the wordmark with the bars mark riding after it like a
// trailing equals sign. Parametric so the hand-tuned alignment (measured at
// wordmark height 64) scales cleanly to any header size.
//
// The pull-left is NEGATIVE SPACING, not an offset. An offset is visual only, so
// the lockup would claim ~14dp of layout width at its right edge that it never
// draws into — phantom width taken straight out of the header's budget for its
// trailing controls.
@Composable
fun TRAQSHeaderLogo(size: Dp = 44.dp, modifier: Modifier = Modifier) {
    Row(
        modifier = modifier,
        horizontalArrangement = Arrangement.spacedBy(-size * (15f / 64f)),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        TRAQSLogo(height = size)
        Box(Modifier.offset(y = -size * (1f / 64f))) {
            TRAQSBarsMark(size = size * (21f / 64f))
        }
    }
}

// Fallback text logo kept for places where an image won't fit
@Composable
fun TRAQSLogoText(fontSize: Int = 24) {
    val c = traQSColors
    Text(
        text = "TRAQS",
        fontSize = fontSize.sp,
        fontWeight = FontWeight.Black,
        style = LocalTextStyle.current.copy(
            brush = Brush.horizontalGradient(listOf(c.accent, parseColor("#2563eb")))
        )
    )
}

// The app's one header: wordmark on the left, trailing actions on the right.
// Mirrors iOS GlassHeader — no centre title, because the tab bar already says
// where you are, and no hamburger, because there is no longer a drawer to open.
//
// TRANSPARENT, deliberately. The shell paints one PageBackground behind every
// tab and the header floats on it; an opaque fill here cut a flat band across
// the top of the liquid canvas.
@Composable
fun TRAQSHeader(actions: @Composable RowScope.() -> Unit = {}) {
    val c = traQSColors
    val logoSize = 56.dp
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .statusBarsPadding()
            // Bottom padding is 2, not 12, and that is not a typo. The wordmark
            // art carries its own transparent margin VERTICALLY as well as on
            // the left, so a 56dp logo box holds roughly 15dp of nothing under
            // the glyph. Twelve more on top of that is what left every page
            // looking like it had a heavy forehead above its title.
            .padding(start = 16.dp, end = 16.dp, top = 12.dp, bottom = 2.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(6.dp)
    ) {
        // Pulled left by the wordmark's own transparent margin, so the visible
        // "t" lands on the same 16dp gutter every PageTitle uses — otherwise the
        // header reads as indented against the title right below it.
        //
        // MEASURED from the PNG's alpha bounding box (the same 2348x1200 asset
        // iOS ships): the glyph starts at x=331, i.e. 14.1% of the width. The
        // 2dp back is the optical nudge iOS applies for the same reason — a big
        // title glyph carries its own side bearing, so a lockup set to the exact
        // gutter reads a touch too far left.
        TRAQSHeaderLogo(
            size = logoSize,
            modifier = Modifier.offset(x = -(logoSize * WORDMARK_ASPECT * 0.141f - 2.dp))
        )
        Spacer(Modifier.weight(1f))
        Row(
            horizontalArrangement = Arrangement.spacedBy(6.dp),
            verticalAlignment = Alignment.CenterVertically
        ) { actions() }
    }
}

// The header for a PUSHED screen — back button on the left, optional actions on
// the right, and the page's own `PageTitle` underneath it.
//
// This exists because those screens used Material3's `TopAppBar`, which reserves
// a fixed 64dp row plus the status-bar inset to hold one small chevron. With the
// title living below it in a `PageTitle`, that bar was 64dp of nothing — the
// "big forehead" every pushed page carried. This row is the same height as the
// tab pages' own header (38dp control + 12 top + 2 bottom), so a pushed screen
// and a tab now start their content on the same line, and the back control is
// the app's glass key rather than a bare Material icon button.
@Composable
fun TRAQSPageBar(
    onBack: () -> Unit,
    backIcon: androidx.compose.ui.graphics.vector.ImageVector = TIcons.ArrowLeft,
    backDescription: String = "Back",
    actions: @Composable RowScope.() -> Unit = {},
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .statusBarsPadding()
            .padding(start = 16.dp, end = 16.dp, top = 12.dp, bottom = 2.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(6.dp)
    ) {
        TRAQSIconBtn(icon = backIcon, contentDescription = backDescription, onClick = onBack)
        Spacer(Modifier.weight(1f))
        Row(
            horizontalArrangement = Arrangement.spacedBy(6.dp),
            verticalAlignment = Alignment.CenterVertically
        ) { actions() }
    }
}

// The header control: a round key in the app's glass language. Mirrors iOS
// HeaderGlassPill — near-solid rather than translucent, because a button has to
// read as an opaque object ON the glass, not as more glass.
@Composable
fun TRAQSIconBtn(
    icon: androidx.compose.ui.graphics.vector.ImageVector,
    contentDescription: String? = null,
    iconColor: Color? = null,
    enabled: Boolean = true,
    onClick: () -> Unit,
) {
    val c = traQSColors
    val tint = iconColor ?: if (enabled) c.text else c.muted
    Box(
        modifier = Modifier
            .size(38.dp)
            .glassControl()
            .clickable(enabled = enabled, onClick = onClick),
        contentAlignment = Alignment.Center
    ) {
        Icon(icon, contentDescription, tint = tint, modifier = Modifier.size(18.dp))
    }
}

// Consistent action bar shown below the header on every main screen
@Composable
fun PageActionBar(
    title: String,
    onAskTRAQS: () -> Unit,
    primaryAction: @Composable RowScope.() -> Unit = {}
) {
    val c = traQSColors
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 16.dp, vertical = 10.dp),
        contentAlignment = Alignment.Center
    ) {
        // Left — Ask TRAQS
        OutlinedButton(
            onClick = onAskTRAQS,
            contentPadding = PaddingValues(horizontal = 12.dp, vertical = 0.dp),
            modifier = Modifier.height(34.dp).align(Alignment.CenterStart),
            shape = RoundedCornerShape(TRadius.xs),
            border = BorderStroke(1.dp, c.accent.copy(alpha = 0.5f)),
            colors = ButtonDefaults.outlinedButtonColors(contentColor = c.accent)
        ) {
            Icon(TIcons.Spark, null, modifier = Modifier.size(14.dp))
            Spacer(Modifier.width(4.dp))
            Text("Ask TRAQS", fontSize = 12.sp, fontWeight = FontWeight.Bold)
        }

        // Center — page title
        Text(
            title,
            fontWeight = FontWeight.Bold,
            fontSize = 20.sp,
            color = c.text,
            modifier = Modifier.align(Alignment.Center)
        )

        // Right — primary action
        Row(
            modifier = Modifier.align(Alignment.CenterEnd),
            horizontalArrangement = Arrangement.spacedBy(8.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            primaryAction()
        }
    }
}
